code/parachain/frame/staking-rewards/src/lib.rs

```diff
@@ -1664,85 +1663,101 @@ fn update_rewards_pool<T: Config>(
-		//
-		// SAFETY(benluelo): Usage of Div::div: Integer division can only fail if rhs == 0, and
-		// reward.reward_rate.amount is known to be non-zero here as per the check at the beginning
-		// of this function
-		let maximum_releasable_periods = total_locked_rewards.div(reward.reward_rate.amount.into());
-
-		let releasable_periods_surpassed =
-			cmp::min(maximum_releasable_periods, periods_surpassed.into());
-
-		// SAFETY: Usage of mul is safe here because newly_accumulated_rewards =
-		//      (    total_locked_rewards            elapsed_time        )
-		//  min ( ------------------------- , -------------------------- )
-		//      ( reward.reward_rate.amount   reward_rate_period_seconds )
-		// If LHS is smaller, this will result in total_locked_rewards, which we know fits in a
-		// u128. If RHS is smaller, this has to be < total_locked_rewards, so it will also fit.
-		let newly_accumulated_rewards =
-			releasable_periods_surpassed.mul(reward.reward_rate.amount.into());
-		let new_total_rewards = newly_accumulated_rewards
-			.safe_add(&reward.total_rewards.into())
-			.map_err(|_| RewardAccumulationCalculationError::ArithmeticError)?;
-
-		// u64::MAX is roughly 584.9 billion years in the future, so saturating at that should be ok
-		let last_updated_timestamp = reward.last_updated_timestamp.saturating_add(
-			reward_rate_period_seconds
-				.get()
-				.saturating_mul(releasable_periods_surpassed.saturated_into::<u64>()),
-		);
-
-		if !newly_accumulated_rewards.is_zero() {
-			T::Assets::release(
-				asset_id,
-				pool_account,
-				newly_accumulated_rewards.into(),
-				false, // not best effort, entire amount must be released
-			)
-			.expect("funds should be available to release based on previous check; qed;");
+	// if no periods have been surpassed, short-circuit
+	let Some(periods_surpassed) = NonZeroU64::new(periods_surpassed) else {
+		return Ok(())
+	};
 
-			reward.total_rewards = new_total_rewards.into();
-			reward.last_updated_timestamp = last_updated_timestamp;
+	let total_locked_rewards: u128 = T::Assets::balance_on_hold(asset_id, pool_account).into();
+
+	// the maximum amount repayable given the reward rate.
+	// i.e. if total locked is 50, and the reward rate is 15, then this would be 3
+	let maximum_releasable_periods =
+		NonZeroU128::new(<u128 as Div<NonZeroU128>>::div(total_locked_rewards, reward_rate_amount))
+			// if the maximum releasable periods is zero, then that means the pot doesn't have
+			// enough in it to fund a single period.
+			.ok_or(RewardAccumulationCalculationError::RewardsPotEmpty)?;
+
+	let releasable_periods_surpassed =
+		cmp::min(maximum_releasable_periods, periods_surpassed.into());
+
+	// SAFETY: Usage of mul is safe here. See the following proof:
+	//
+	// ```plaintext
+	//                                    (  total_locked_rewards           elapsed_time        )
+	// releasable_periods_surpassed = min ( ---------------------- , -------------------------- )
```

Comment:
I would just define some symbols to represent each value when things become this long. Not easy to read as of now.

Reply:
I was going to define symbols, but I wanted to show explicitly what each part was, to more easily explain the proof. I agree that it is somewhat unwieldy, I'll see what I can do